File: content/problems/2767-partition-string-into-minimum-beautiful-substrings/solutions/prefix-dp-with-powers-of-five/solution.cpp
class Solution {
    public:int minimumBeautifulSubstrings(string s){
        unordered_set<long long> powers;
        long long limit=1LL<<s.size();
        for(long long value=1;value<limit;value*=5)powers.insert(value);
        const int INF=1000000;
        vector<int> dp(s.size()+1,INF);
        dp[0]=0;
        for(int start=0;start<(int)s.size();++start)if(dp[start]<INF&&s[start]!='0'){
            long long value=0;
            for(int end=start;end<(int)s.size();++end){
                value=value*2+s[end]-'0';
                if(powers.count(value))dp[end+1]=min(dp[end+1],dp[start]+1);
            }
        }
        return dp.back()==INF?-1:dp.back();
    }
};

File: content/problems/2767-partition-string-into-minimum-beautiful-substrings/solutions/prefix-dp-with-powers-of-five/Solution.java
class Solution {
    public int minimumBeautifulSubstrings(String s){
        Set<Long> powers=new HashSet<>();
        long limit=1L<<s.length();
        for(long value=1;value<limit;value*=5)powers.add(value);
        int inf=1000000,n=s.length();
        int[] dp=new int[n+1];
        Arrays.fill(dp,inf);
        dp[0]=0;
        for(int start=0;start<n;start++)if(dp[start]<inf&&s.charAt(start)!='0'){
            long value=0;
            for(int end=start;end<n;end++){
                value=value*2+s.charAt(end)-'0';
                if(powers.contains(value))dp[end+1]=Math.min(dp[end+1],dp[start]+1);
            }
        }
        return dp[n]==inf?-1:dp[n];
    }
}
